starknet_py/cairo/serialization/_context.py

```diff
@@ -0,0 +1,130 @@
+from __future__ import annotations
+
+from abc import ABC
+from contextlib import contextmanager
+from typing import List, Generator, ContextManager, Any
+
+from starknet_py.cairo.serialization._calldata_reader import (
+    OutOfBoundsError,
+    CalldataReader,
+    CairoData,
+)
+from starknet_py.cairo.serialization.errors import (
+    InvalidValueException,
+    InvalidTypeException,
+)
+
+
+class Context(ABC):
+    """
+    Holds information about context when (de)serializing data. This is needed to inform what and where went
+    wrong during processing. Every separate (de)serialization should have its own context.
+    """
+
+    _namespace_stack: List[str]
+
+    def __init__(self):
+        self._namespace_stack = []
+
+    @property
+    def current_entity(self):
+        """
+        Name of currently processed entity.
+
+        :return: transformed path.
+        """
+        return ".".join(self._namespace_stack)
+
+    @contextmanager
+    def push_entity(self, name: str) -> Generator:
+        """
+        Manager used for maintaining information about names of (de)serialized types. Will wrap some errors with
+        custom errors, adding information about the context.
+
+        :param name: name of (de)serialized entity.
+        """
+        self._namespace_stack.append(name)
+        yield
+        self._namespace_stack.pop()
```

Comment:
It makes sure that if everything works properly the value will be removed from the stack. The error is only caught at the end, so we want the stack to be filled in case of error.

Reply:
I'll add a comment

Reply:
So context have both push_entity and a _wrap_method that is also a context manager. What is the interaction between this two? Why there have to be two context managers?

Reply:
`push_entity` is pushing and popping current entity. 
`_wrap_errors` wraps errors thrown. `_wrap_errors` was part of push_entity before, but the problem was when processing top level entity - there was no entity pushed and therefore error was not wrapped. 

I added `.create` methods to create contexts that automatically ensures all errors are handled. Since this is in place there is no point in having error handling in push_entity.